File: backend/app/schemas/dev.py
"""개발자 모드 API 스키마 (R7-03) — /api/v1/dev 요청·응답.

요청 스키마는 전부 extra='forbid' — dev 도구는 오타(예: cloud→clouds)를 조용히
무시하면 "조작이 안 먹었는데 200"이 되므로 명시적 422로 거부한다(onboarding 관례).
검증 규칙 중 env로 튜닝되는 상한(구름 CLOUD_MAX)은 라우터가 런타임 검증한다 —
스키마 리터럴로 굳히면 env 튜닝과 드리프트가 나기 때문.
"""
from datetime import date
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field


class DevAbilityOut(BaseModel):
    """GET /dev/state의 abilities 원소 — user_concept_ability 원값 그대로."""

    concept_tag: str
    theta: float
    theta_se: float
    num_responses: int


class DevState(BaseModel):
    """GET /dev/state 응답 — 프론트는 이 엔드포인트 200 여부로 패널 노출을 결정."""

    dev_mode: bool = True
    abilities: list[DevAbilityOut]
    overall_theta: float | None
    target_level_group: str
    unlock_floor: int
    clouds: int
    streak_count: int
    placement_done: bool
    weak_tags: list[str]


class DevResetResult(BaseModel):
    """POST /dev/reset-me 응답."""

    reset: bool = True


class DevThetaAbilityIn(BaseModel):
    """POST /dev/theta의 abilities 원소 — theta_se는 서버 기본(0.3)."""

    model_config = ConfigDict(extra="forbid")

    concept_tag: str
    theta: float
    num_responses: int = Field(default=1, ge=0)


class DevThetaRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    abilities: list[DevThetaAbilityIn] = Field(min_length=1)


class DevThetaResult(BaseModel):
    updated: int


class DevPlacementRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    action: Literal["reset", "complete"]


class DevPlacementResult(BaseModel):
    placement_done: bool


class DevCloudsRequest(BaseModel):
    """상한(CLOUD_MAX)은 env 튜닝값이라 라우터가 런타임 검증한다."""

    model_config = ConfigDict(extra="forbid")

    clouds: int = Field(ge=0)


class DevCloudsResult(BaseModel):
    clouds: int
    max: int


class DevCurriculumRequest(BaseModel):
    """unit_slug·crowns는 action='crown'일 때만 필수 — 라우터가 검증."""

    model_config = ConfigDict(extra="forbid")

    action: Literal["unlock_all", "crown", "reset"]
    unit_slug: str | None = None
    crowns: int | None = Field(default=None, ge=0)


class DevCurriculumResult(BaseModel):
    """affected: 조작이 닿은 유닛 진도 행 수(unlock_all=유닛 수, crown=1, reset=삭제 행)."""

    action: str
    affected: int


class DevStreakRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    streak_count: int = Field(ge=0)
    last_login_days_ago: int = Field(default=0, ge=0)


class DevStreakResult(BaseModel):
    streak_count: int
    last_login_date: date

File: frontend/src/api/dev.js
import client from './client';

/**
 * Dev API (스프린트 R7-03 계약 — /api/v1/dev)
 * 개발자 모드 전용 인스펙터·조작 엔드포인트. DEV_MODE가 꺼진 서버는 전 경로가
 * 404를 돌려주며, 프론트는 GET /dev/state 404를 노출 게이트로 사용한다
 * (404면 DevPanel 자체를 렌더하지 않음 — 프론트 env 불필요).
 *
 * POST 응답 본문 형태는 백엔드 병렬 구현 중이라 미확정 — 프론트는 응답 본문에
 * 의존하지 않고, 조작 후 react-query invalidate로 화면을 갱신한다.
 */

// GET /dev/state — 개발자 모드 상태 인스펙터.
//   → {dev_mode: true,
//      abilities: [{concept_tag, theta, theta_se, num_responses}],
//      overall_theta, target_level_group, unlock_floor,
//      clouds, streak_count, placement_done, weak_tags}
//   DEV_MODE 꺼짐 → 404 (패널 미렌더 게이트).
export async function fetchDevState() {
  const res = await client.get('/dev/state');
  return res.data;
}

// POST /dev/reset-me {reset: true} — 내 계정 진행 전체 초기화(진도·θ·배치·자원).
export async function resetMe() {
  const res = await client.post('/dev/reset-me', { reset: true });
  return res.data;
}

// POST /dev/theta {abilities: [{concept_tag, theta, num_responses?}]} — θ 직접 설정.
export async function setTheta(abilities) {
  const res = await client.post('/dev/theta', { abilities });
  return res.data;
}

// POST /dev/placement {action: "reset"|"complete"} — 배치고사 초기화/즉시완료.
export async function setPlacement(action) {
  const res = await client.post('/dev/placement', { action });
  return res.data;
}

// POST /dev/clouds {clouds} — 구름 에너지 잔량 직접 설정(0..최대).
export async function setClouds(clouds) {
  const res = await client.post('/dev/clouds', { clouds });
  return res.data;
}

// POST /dev/curriculum {action: "unlock_all"|"crown"|"reset", unit_slug?, crowns?}
//   unlock_all: 전 유닛 잠금 해제 / crown: 특정 유닛 왕관 설정 / reset: 진도 초기화.
export async function setCurriculum({ action, unitSlug, crowns } = {}) {
  const body = { action };
  if (unitSlug != null) body.unit_slug = unitSlug;
  if (crowns != null) body.crowns = crowns;
  const res = await client.post('/dev/curriculum', body);
  return res.data;
}

// POST /dev/streak {streak_count, last_login_days_ago?} — 스트릭 상태 설정.
export async function setStreak({ streakCount, lastLoginDaysAgo } = {}) {
  const body = { streak_count: streakCount };
  if (lastLoginDaysAgo != null) body.last_login_days_ago = lastLoginDaysAgo;
  const res = await client.post('/dev/streak', body);
  return res.data;
}
